fdbcli/DebugCommands.actor.cpp

```diff
@@ -294,78 +302,97 @@ ACTOR Future<bool> checkallCommandActor(Database cx, std::vector<StringRef> toke
 			    keyServerPromise.getFuture().get();
-			state Key lastEnd; // this is to hold the Key having the value of next begin
-			for (i = 0; i < keyServers.size(); i++) { // for each key range
-				state KeyRange range = keyServers[i].first;
-				range = range & toCheck;
-				if (range.empty()) {
+
+			// Checking each shard
+			state int i = 0;
+			for (; i < keyServers.size(); i++) { // for each shard
+				state KeyRange rangeToCheck = keyServers[i].first;
+				rangeToCheck = rangeToCheck & inputRange; // Only check the shard part within the inputRange
+				if (rangeToCheck.empty()) {
 					continue;
 				}
 				const auto& servers = keyServers[i].second;
-				begin = firstGreaterOrEqual(range.begin);
-				end = firstGreaterOrEqual(range.end);
-				printf("Key range: %s\n", printable(range).c_str());
+				state Key beginKeyToCheck = rangeToCheck.begin;
+				printf("Key range to check: %s\n", printable(rangeToCheck).c_str());
 				for (const auto& server : servers) {
 					printf("  %s\n", server.address().toString().c_str());
 				}
 				state std::vector<Future<ErrorOr<GetKeyValuesReply>>> replies;
 				state bool hasMore = true;
 				state int round = 0;
+				state Version version;
 				while (hasMore) {
 					wait(store(version, getVersion(cx)));
 					replies.clear();
-					printf("round %d, begin Key: %s\n", round, printable(begin.toString()).c_str());
+					printf("Round %d: %s - %s\n",
+					       round,
+					       printable(beginKeyToCheck.toString()).c_str(),
+					       printable(rangeToCheck.end.toString()).c_str());
 					for (const auto& s : keyServers[i].second) { // for each storage server
 						GetKeyValuesRequest req;
-						req.begin = begin;
-						req.end = end;
+						req.begin = firstGreaterOrEqual(beginKeyToCheck);
+						req.end = firstGreaterOrEqual(rangeToCheck.end);
 						req.limit = CLIENT_KNOBS->KRM_GET_RANGE_LIMIT;
 						req.limitBytes = CLIENT_KNOBS->KRM_GET_RANGE_LIMIT_BYTES;
-						req.version = version;
+						req.version = version; // all replica should read at the same version
 						req.tags = TagSet();
 						replies.push_back(s.getKeyValues.getReplyUnlessFailedFor(req, 2, 0));
 					}
-					// printf("waiting for %lu replies at version: %ld\n", keyServers[i].second.size(), version);
 					wait(waitForAll(replies));
-					// if there are more results, continue checking in the same shard
-					auto p = checkResults(version, keyServers[i].second, replies, begin, end, lastEnd);
-					bool allSame = p.first;
-					hasMore = p.second;
-					printf("AllSame %d, hasMore %d, checkAll %d\n", allSame, hasMore, checkAll);
+					// checkResults keeps invariants:
+					// (1) hasMore = true if any server has more data not read yet
+					// (2) nextBeginKey is the minimal key returned from all servers
+					// (3) checkResults reports inconsistency of keys only before the nextBeginKey if hasMore=true
+					// Therefore, whether to proceed to the next round depends on hasMore
```

Comment:
These repeat the comments in the definition above, could be removed.

Reply:
Nice catch!